packages/editor/src/optimizer/common.ts

```diff
@@ -0,0 +1,614 @@
+import {
+  Canvas,
+  collectLabels,
+  LabelCache,
+  LabelData,
+  LabelMeasurements,
+  mathjaxInit,
+  Num,
+  PenroseError,
+  PenroseWarning,
+  Shape,
+  State,
+} from "@penrose/core";
+import consola from "consola";
+import { Result } from "true-myth";
+
+// Config
+
+export const logLevel = (consola as any).LogLevel.Warn;
+
+// Basic types
+
+/** Unique ID assigned to every request/response pair */
+export type MessageID = number;
+
+/**
+ * Unique ID assigned to every new worker, which each handle one diagrams
+ * compilation, optimization, interaction, etc.
+ */
+export type DiagramID = number;
+
+/**
+ * Unique ID assigned to every sequence of optimizer steps. Only necessarily
+ * unique within a diagram.
+ */
+export type StepSequenceID = number;
+
+/**
+ * History is a DAG of `StepSequences`, which are a series of optimizer
+ * steps, stages, and states. Each of these layout sequences has an id, and
+ * the steps are indexed by order.
+ */
+export type HistoryLoc = {
+  sequenceId: StepSequenceID;
+  step: number;
+};
+
+/**
+ * Array of stage name and optimizer steps per stage.
+ */
+export type LayoutStats = {
+  name: string;
+  steps: number;
+  cumulativeSteps: number;
+}[];
+
+/**
+ * Current state of a sequence of optimizer states:
+ *  - `"Pending"`: optimizer has not finished (ongoing, interrupted). Could
+ *    or will continue.
+ *  - `"Done"`: optimizer finished normally
+ *  - `OptimizerError`: optimizer errored and cannot continue.
+ */
+export type StepSequenceState = "Pending" | "Done" | OptimizationError;
```

Comment:
Is there a reason you don't do this instead?

```typescript
export type StepSequenceState =
  | { tag: "Pending" }
  | { tag: "Done" }
  | OptimizationError;
```

Seems like this would be more convenient/consistent, because then you can just do `switch (stepSequenceState.tag)`.

Reply:
This is a better idea, I'll change that.